File: beanmachine/graph/graph_test.py
# Copyright (c) Facebook, Inc. and its affiliates.
import unittest

import torch
from beanmachine import graph


class TestBayesNet(unittest.TestCase):
    def test_simple_dep(self):
        g = graph.Graph()
        c1 = g.add_constant(torch.FloatTensor([0.8, 0.2]))
        d1 = g.add_distribution(
            graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, [c1]
        )
        g.add_operator(graph.OperatorType.SAMPLE, [d1])

    def test_tabular(self):
        g = graph.Graph()
        c1 = g.add_constant(torch.FloatTensor([0.8, 0.2]))

        # negative test
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, []
            )
        self.assertTrue(
            "Tabular distribution first arg must be tensor" in str(cm.exception)
        )

        g = graph.Graph()
        c1 = g.add_constant(torch.FloatTensor([0.8, 0.2]))
        var1 = g.add_operator(
            graph.OperatorType.SAMPLE,
            [
                g.add_distribution(
                    graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, [c1]
                )
            ],
        )
        var2 = g.add_operator(
            graph.OperatorType.SAMPLE,
            [
                g.add_distribution(
                    graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, [c1]
                )
            ],
        )

        # since the following has two parents it must have a tabular dist with
        # 3 dimensions in the tensor
        with self.assertRaises(ValueError) as cm:
            g.add_operator(
                graph.OperatorType.SAMPLE,
                [
                    g.add_distribution(
                        graph.DistributionType.TABULAR,
                        graph.AtomicType.BOOLEAN,
                        [c1, var1, var2],
                    )
                ],
            )
        self.assertTrue("expected 3 dims got 1" in str(cm.exception))

        c2 = g.add_constant(torch.FloatTensor([[0.6, 0.4], [0.99, 0.01]]))
        g.add_distribution(
            graph.DistributionType.TABULAR,
            graph.AtomicType.BOOLEAN,
            [c2, g.add_constant(True)],
        )

        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.TABULAR,
                graph.AtomicType.BOOLEAN,
                [c2, g.add_constant(torch.tensor(1))],
            )
        self.assertTrue("only supports boolean parents" in str(cm.exception))

        c3 = g.add_constant(torch.FloatTensor([1.1, -0.1]))
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, [c3]
            )
        self.assertTrue("must be positive" in str(cm.exception))

        c4 = g.add_constant(torch.FloatTensor([0.9, 0.0999]))
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, [c4]
            )
        self.assertTrue("must add to 1" in str(cm.exception))

    def test_bernoulli(self):
        g = graph.Graph()
        c1 = g.add_constant_probability(1.0)
        c2 = g.add_constant_probability(0.8)

        # negative tests on number of parents
        # 0 parents not allowed
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, []
            )
        self.assertTrue(
            "Bernoulli distribution must have exactly one parent" in str(cm.exception)
        )
        # 2 parents not allowed
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [c1, c2]
            )
        self.assertTrue(
            "Bernoulli distribution must have exactly one parent" in str(cm.exception)
        )

        # 1 parent is OK
        d1 = g.add_distribution(
            graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [c1]
        )

        # negative test on type of parent
        c3 = g.add_constant(torch.scalar_tensor(1, dtype=torch.int))
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [c3]
            )
        self.assertTrue("must be a probability" in str(cm.exception))

        # negative test on value of parent
        with self.assertRaises(ValueError) as cm:
            g.add_constant_probability(1.1)
        self.assertTrue("must be between 0 and 1" in str(cm.exception))

        v1 = g.add_operator(graph.OperatorType.SAMPLE, [d1])
        g.query(v1)
        samples = g.infer(1)
        self.assertTrue(samples[0][0].type == graph.AtomicType.BOOLEAN)
        self.assertTrue(samples[0][0].bool)
        means = g.infer_mean(1)
        self.assertEqual(len(means), 1, "exactly one node queried")

    def test_beta(self):
        g = graph.Graph()
        c1 = g.add_constant(1.1)
        c2 = g.add_constant(5.0)
        # negative tests on number of parents
        # 0 parents not allowed
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.BETA, graph.AtomicType.PROBABILITY, []
            )
        self.assertTrue(
            "Beta distribution must have exactly two parents" in str(cm.exception)
        )
        # 1 parent not allowed
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.BETA, graph.AtomicType.PROBABILITY, [c1]
            )
        self.assertTrue(
            "Beta distribution must have exactly two parents" in str(cm.exception)
        )
        # negative test on type of parent
        c3 = g.add_constant(True)
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.BETA, graph.AtomicType.PROBABILITY, [c3, c3]
            )
        self.assertTrue("must be real-valued" in str(cm.exception))
        # negative test on sample type
        with self.assertRaises(ValueError) as cm:
            g.add_distribution(
                graph.DistributionType.BETA, graph.AtomicType.REAL, [c1, c2]
            )
        self.assertTrue("Beta produces probability samples" in str(cm.exception))
        # 2 real-valued parents with probability sample type are OK
        d1 = g.add_distribution(
            graph.DistributionType.BETA, graph.AtomicType.PROBABILITY, [c1, c2]
        )
        # now let's draw some samples from the Beta distribution
        v1 = g.add_operator(graph.OperatorType.SAMPLE, [d1])
        g.query(v1)
        samples = g.infer(1, graph.InferenceType.REJECTION)
        self.assertTrue(samples[0][0].type == graph.AtomicType.PROBABILITY)
        self.assertTrue(samples[0][0].probability > 0 and samples[0][0].probability < 1)
        means = g.infer_mean(10000, graph.InferenceType.REJECTION)
        self.assertAlmostEqual(means[0], 1.1 / (1.1 + 5.0), 2, "beta mean")

    def test_binomial(self):
        g = graph.Graph()
        c1 = g.add_constant(10)
        c2 = g.add_constant_probability(0.55)
        d1 = g.add_distribution(
            graph.DistributionType.BINOMIAL, graph.AtomicType.NATURAL, [c1, c2]
        )
        v1 = g.add_operator(graph.OperatorType.SAMPLE, [d1])
        g.query(v1)
        samples = g.infer(1, graph.InferenceType.REJECTION)
        self.assertTrue(samples[0][0].type == graph.AtomicType.NATURAL)
        self.assertTrue(samples[0][0].natural <= 10)
        means = g.infer_mean(10000, graph.InferenceType.REJECTION)
        self.assertTrue(means[0] > 5 and means[0] < 6)

    def _create_graph(self):
        g = graph.Graph()
        c1 = g.add_constant(torch.FloatTensor([0.8, 0.2]))
        c2 = g.add_constant(torch.FloatTensor([[0.6, 0.4], [0.99, 0.01]]))
        c3 = g.add_constant(
            torch.FloatTensor([[[1, 0], [0.2, 0.8]], [[0.1, 0.9], [0.01, 0.99]]])
        )
        Rain = g.add_operator(
            graph.OperatorType.SAMPLE,
            [
                g.add_distribution(
                    graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, [c1]
                )
            ],
        )
        Sprinkler = g.add_operator(
            graph.OperatorType.SAMPLE,
            [
                g.add_distribution(
                    graph.DistributionType.TABULAR, graph.AtomicType.BOOLEAN, [c2, Rain]
                )
            ],
        )
        GrassWet = g.add_operator(
            graph.OperatorType.SAMPLE,
            [
                g.add_distribution(
                    graph.DistributionType.TABULAR,
                    graph.AtomicType.BOOLEAN,
                    [c3, Sprinkler, Rain],
                )
            ],
        )
        return g, Rain, Sprinkler, GrassWet

    def test_query(self):
        g, Rain, Sprinkler, GrassWet = self._create_graph()
        g.query(Rain)
        g.query(Sprinkler)
        g.query(GrassWet)
        g.infer(1)

    def test_observe(self):
        g, Rain, Sprinkler, GrassWet = self._create_graph()
        g.observe(GrassWet, True)
        with self.assertRaises(ValueError) as cm:
            g.observe(GrassWet, True)
        self.assertTrue("duplicate observe for node" in str(cm.exception))

        g = graph.Graph()
        c1 = g.add_constant_probability(1.0)
        c2 = g.add_constant_probability(0.5)
        o1 = g.add_operator(graph.OperatorType.MULTIPLY, [c1, c2])
        d1 = g.add_distribution(
            graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [o1]
        )
        o2 = g.add_operator(graph.OperatorType.SAMPLE, [d1])
        with self.assertRaises(ValueError) as cm:
            g.observe(o1, True)
        self.assertTrue("only sample nodes may be observed" in str(cm.exception))
        g.observe(o2, True)  # ok to observe this node

    def test_inference(self):
        g, Rain, Sprinkler, GrassWet = self._create_graph()
        g.observe(GrassWet, True)
        g.query(Rain)
        g.query(GrassWet)
        with self.assertRaises(ValueError) as cm:
            g.query(Rain)
        self.assertTrue("duplicate query for node" in str(cm.exception))
        samples = g.infer(1)
        self.assertTrue(len(samples) == 1)
        # since we have observed grass wet is true the query should be true
        self.assertTrue(samples[0][1].type == graph.AtomicType.BOOLEAN)
        self.assertTrue(samples[0][1].bool)

    def test_infer_mean(self):
        g = graph.Graph()
        c1 = g.add_constant_probability(1.0)
        op1 = g.add_operator(graph.OperatorType.MULTIPLY, [c1, c1])
        d1 = g.add_distribution(
            graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [op1]
        )
        op2 = g.add_operator(graph.OperatorType.SAMPLE, [d1])
        g.query(op1)
        g.query(op2)
        means = g.infer_mean(100)
        self.assertAlmostEqual(means[0], 1.0)
        self.assertAlmostEqual(means[1], 1.0)
        # negative test, don't support aggregating tensors
        c2 = g.add_constant(torch.tensor(0.5))
        op2 = g.add_operator(graph.OperatorType.ADD, [c2, c2])
        g.query(op2)
        with self.assertRaises(RuntimeError):
            g.infer_mean(100)
        g.infer(100)  # infer should be fine though

File: beanmachine/graph/graph_test.cpp
// Copyright (c) Facebook, Inc. and its affiliates.
#include <array>
#include <tuple>

#include <gtest/gtest.h>
#include <torch/torch.h>

#include "beanmachine/graph/graph.h"

using namespace beanmachine;

TEST(testgraph, infer_arithmetic) {
  graph::Graph g;
  uint c1 = g.add_constant_probability(0.1);
  uint d1 = g.add_distribution(
      graph::DistributionType::BERNOULLI,
      graph::AtomicType::BOOLEAN,
      std::vector<uint>({c1}));
  uint o1 =
      g.add_operator(graph::OperatorType::SAMPLE, std::vector<uint>({d1}));
  uint o2 =
      g.add_operator(graph::OperatorType::TO_POS_REAL, std::vector<uint>({o1}));
  uint c2 = g.add_constant_pos_real(0.8);
  uint o3 = g.add_operator(
      graph::OperatorType::MULTIPLY, std::vector<uint>({c2, o2}));
  uint c3 = g.add_constant_pos_real(0.1);
  uint o4 =
      g.add_operator(graph::OperatorType::ADD, std::vector<uint>({c3, o3}));
  uint d2 = g.add_distribution(
      graph::DistributionType::BERNOULLI_NOISY_OR,
      graph::AtomicType::BOOLEAN,
      std::vector<uint>({o4}));
  uint o5 =
      g.add_operator(graph::OperatorType::SAMPLE, std::vector<uint>({d2}));
  // P(o5|o1=T) = 1 - exp(-.9)=0.5934 and P(o5|o1=F) = 1-exp(-.1)=0.09516
  // Since P(o1=T)=0.1 and P(o1=F)=0.9. Therefore P(o5=T,o1=T) = 0.05934,
  // P(o5=T,o1=F) = 0.08564 and P(o1=T | o5=T) = 0.4093
  g.observe(o5, true);
  g.query(o1);
  std::vector<std::vector<graph::AtomicValue>> samples =
      g.infer(100, graph::InferenceType::GIBBS);
  int sum = 0;
  for (const auto& sample : samples) {
    const auto& s = sample.front();
    ASSERT_EQ(s.type, graph::AtomicType::BOOLEAN);
    sum += s._bool ? 1 : 0;
  }
  // less than 1 in million odds of failing these tests with correct infer
  EXPECT_LT(sum, 65);
  EXPECT_GT(sum, 15);
  // infer_mean should give almost exactly the same answer
  std::vector<double> means = g.infer_mean(100, graph::InferenceType::GIBBS);
  EXPECT_TRUE(std::abs(sum - int(means[0] * 100)) <= 1);
  // repeat the test with rejection sampling
  std::vector<std::vector<graph::AtomicValue>> samples2 =
      g.infer(100, graph::InferenceType::REJECTION);
  sum = 0;
  for (const auto& sample : samples2) {
    const auto& s = sample.front();
    ASSERT_EQ(s.type, graph::AtomicType::BOOLEAN);
    sum += s._bool ? 1 : 0;
  }
  EXPECT_LT(sum, 65);
  EXPECT_GT(sum, 15);
  // infer_mean should give the same answer
  std::vector<double> means2 =
      g.infer_mean(100, graph::InferenceType::REJECTION);
  EXPECT_TRUE(std::abs(sum - int(means2[0] * 100)) <= 1);
}

TEST(testgraph, infer_bn) {
  graph::Graph g;
  // classic sprinkler BN, see the diagram here:
  // https://upload.wikimedia.org/wikipedia/commons/0/0e/SimpleBayesNet.svg
  uint c1 = g.add_constant(torch::from_blob((float[]){.8, .2}, {2}));
  uint d1 = g.add_distribution(
      graph::DistributionType::TABULAR,
      graph::AtomicType::BOOLEAN,
      std::vector<uint>({c1}));
  uint RAIN =
      g.add_operator(graph::OperatorType::SAMPLE, std::vector<uint>({d1}));
  uint c2 =
      g.add_constant(torch::from_blob((float[]){.6, .4, .99, .01}, {2, 2}));
  uint d2 = g.add_distribution(
      graph::DistributionType::TABULAR,
      graph::AtomicType::BOOLEAN,
      std::vector<uint>({c2, RAIN}));
  uint SPRINKLER =
      g.add_operator(graph::OperatorType::SAMPLE, std::vector<uint>({d2}));
  uint c3 = g.add_constant(
      torch::from_blob((float[]){1, 0, .2, .8, .1, .9, .01, .99}, {2, 2, 2}));
  uint d3 = g.add_distribution(
      graph::DistributionType::TABULAR,
      graph::AtomicType::BOOLEAN,
      std::vector<uint>({c3, SPRINKLER, RAIN}));
  uint GRASSWET =
      g.add_operator(graph::OperatorType::SAMPLE, std::vector<uint>({d3}));
  g.observe(GRASSWET, true);
  g.query(RAIN);
  const std::vector<std::vector<graph::AtomicValue>>& samples =
      g.infer(100, graph::InferenceType::REJECTION);
  ASSERT_EQ(samples.size(), 100);
  uint sum = 0;
  for (const auto& sample : samples) {
    const auto& s = sample.front();
    sum += s._bool ? 1 : 0;
  }
  // true probability is approx .3577
  // less than 1 in million odds of failing these tests with correct infer
  EXPECT_LT(sum, 60);
  EXPECT_GT(sum, 10);
}
